[IBM] Update HTTPS docs with KFServing config steps (#2409)
kubeflow/website: content/en/docs/ibm/deploy/authentication.md

of the column `Hostname` in step 4:

    ```YAML
    apiVersion: networking.istio.io/v1alpha3
    kind: Gateway
    metadata:
      name: kubeflow-gateway
      namespace: kubeflow
    spec:
      selector:
        istio: ingressgateway
      servers:
      - hosts:
        - '<hostname>'
        port:
          name: https
          number: 443
          protocol: HTTPS
        tls:
          mode: SIMPLE
          privateKey: /etc/istio/ingressgateway-certs/tls.key
          serverCertificate: /etc/istio/ingressgateway-certs/tls.crt
    ```

8. Verify that the traffic is routed via HTTPS by using the value of
above-mentioned `Hostname` in your browser. It should redirect traffic from an
HTTP address to HTTPS address automatically.

**Note**: The certificates for the NLB DNS host secret expire every **90** days.
The secret in the `default` namespace is automatically renewed by IBM Cloud
Kubernetes Service 37 days before it expires. After this secret is updated, you
must manually copy it to the `istio-ingressgateway-certs` secret by repeating
commands in step 5 and 6.

## Optional - KFServing configuration

> With this HTTPS setup, additional changes are needed to get KFServing to work.

Reviewer: Suggestion:

- Change to active voice from passive (to help the users by directly addressing):

```suggestion
With this HTTPS setup, you need to make additional changes to get KFServing to work.
```

1. First, update the Knative domain that is used for the KFServing routes to the
hostname that you used when updating `kubeflow-gateway`.

    ```shell
    kubectl edit configmap config-domain -n knative-serving
    ```

    This will open your default text editor, and you will see something like:

    ```YAML
    apiVersion: v1
    data:
      _example: |
        ################################
        #                              #
        #    EXAMPLE CONFIGURATION     #
        #                              #
        ################################
        # ...
        example.com: |
    kind: ConfigMap
    ...
    ```

    Add a line above the `_example` key with your hostname as the key and an empty string value.
    Be sure to update `<hostname>`:

    ```YAML
    apiVersion: v1
    data:
      <hostname>: ""
      _example: |
       ...
    kind: ConfigMap
    ...
    ```

>     After you save and exit, routes for your InferenceServices will use this new domain.

Reviewer: Suggestion:

- Instruct the user to save and exit explicitly:

```suggestion
    Then, save and exit. The routes for your InferenceServices will start using this new domain.
```

WDYT?

> 2. Since the certificates provided by IBM only allow for a single level of domain name added

Reviewer: Suggestion:

- Add "Cloud" to "IBM"

```suggestion
2. Since the certificates provided by IBM Cloud only allow for a single level of domain name added
```

        # ...
    kind: ConfigMap
    ...
    ```

    Add a line above the `_example` key with the `domainTemplate` key like the following:

    ```YAML
    apiVersion: v1
    data:
      domainTemplate: "{{.Name}}-{{.Namespace}}.{{.Domain}}"
      _example: |
       ...
    kind: ConfigMap
    ...
    ```

    Save and exit. The default template uses a two-level subdomain (i.e. `{{.Name}}.{{.Namespace}}.{{.Domain}}`),
    so this just adjusts it to use one.

3. Adjust `kubeflow-gateway` one more time, adding a wildcard host in the HTTPS `hosts` section.

    ```shell
    kubectl edit gateway kubeflow-gateway -n kubeflow
    ```

    This will open your default text editor, but you can also optionally edit `kubeflow-gateway.yaml` file
    you created previously. Here, just add another entry to the HTTPS `hosts` list containing your hostname prepended
    with a `*.` so that the Knative subdomains are properly routed.

    ```YAML
    apiVersion: networking.istio.io/v1alpha3
    kind: Gateway
    metadata:
      name: kubeflow-gateway
      namespace: kubeflow
    spec:
      selector:
        istio: ingressgateway
      servers:
      - hosts:
        - '<hostname>'

>         - '*.<hostname>'

Reviewer 2: Have you tried the sudomain works with a deployed Kubeflow with authentication? I assume AppID will redirect the request to the configured FQDN without the subdomain.

Author: Hi @shawnzhu, thanks for the review. I can certainly split this PR up to separate the scopes.

I have tested, and something like the following curl command works:

```
curl -v https://sklearn-iris-bar-p-test.kfp-dev-442dbba0442be6c8c50f31ed96b00601-0001.sjc03.containers.appdomain.cloud/v1/models/sklearn-iris-bar:predict -d '{"instances": [[6.8,  2.8,  4.8,  1.4],[6.0,  3.4,  4.5,  1.6]]}' -H "Authorization: Bearer eyJhb..."

{"predictions": [1, 1]}
```

Without the Authorization header, I get a HTTP 302, and then what I am assuming is a redirect to the auth UI.

KFserving does not have it's own gateway, it either uses Knative's gateway (`knative-ingress-gateway`) if standalone or `kubeflow-gateway` if KFServing is installed with the rest of Kubeflow via the manifests.

Reviewer 2: > Without the Authorization header, I get a HTTP 302, and then what I am assuming is a redirect to the auth UI

that's what I mentioned above. it would (as designed) pass through the envoyfilter listens at gateway level when providing an authorization header. (which authservice image did you use? it would be very helpful to share which kfdef you've used to deploy this kubeflow).

Author: The KFDef used was this one: https://raw.githubusercontent.com/kubeflow/manifests/v1.2-branch/kfdef/kfctl_ibm_multi_user.v1.2.0.yaml 

I did get Kfserving predict requests working just using HTTP and without the need for any auth headers (thanks to @yhwang) by adding the following to `kubeflow-gateway`.

```
- hosts:
  - '*.<hostname>'
  port:
      name: http-kfserving
      number: 80
      protocol: HTTP
```

Then updating the envoyfilter (authn-filter) to just listen on port 443:
```
    listenerMatch:
      portNumber: 443
      listenerType: GATEWAY 
```
However, I am unsure if these adjustments should be included here.

Reviewer 2: @pvaneck Thanks! that works!

    ```

    This will open your default text editor, but you can also optionally edit `kubeflow-gateway.yaml` file
    you created previously. Here, just add another entry to the HTTPS `hosts` list containing your hostname prepended
    with a `*.` so that the Knative subdomains are properly routed.

    ```YAML
    apiVersion: networking.istio.io/v1alpha3
    kind: Gateway
    metadata:
      name: kubeflow-gateway
      namespace: kubeflow
    spec:
      selector:
        istio: ingressgateway
      servers:
      - hosts:
        - '<hostname>'
        - '*.<hostname>'
        port:
          name: https
          number: 443
          protocol: HTTPS
        tls:
          mode: SIMPLE
          privateKey: /etc/istio/ingressgateway-certs/tls.key
          serverCertificate: /etc/istio/ingressgateway-certs/tls.crt
    ```

    Save and exit.

After these adjustments, InferenceServices should now be reachable via HTTPS. However, if you are using IBM Cloud AppID for
authentication and you want to predict externally, you will need to generate a token as
[described here](https://cloud.ibm.com/docs/appid?topic=appid-custom-auth#generating-jwts). This token can be used as a header

> (e.g. `"Authorization: Bearer eyJh..."`) in your request.

Reviewer 2: Do you mind adding an example about how to use this token to access kfserving? I'm curious which component is validating this JWT token from AppID since the envoyfilter won't allow `Authorization` header go through.

Author: Ah sorry, good catch. I didn't realize that the environment I was using was further customized with the envoyFilter including the "Authorization" pattern in the allowed_headers, and a newer authservice image (http://gcr.io/arrikto/oidc-authservice:087a340) was being used (one that includes a recent PR https://github.com/arrikto/oidc-authservice/commit/087a340cb454766891fdd3092955954c06d154ac).

In that case, I think I may just leave this JWT section out for now.

Reviewer 2: @pvaneck that's what I guessed. Yes, it will work with the latest oidc-authservice which depends on https://github.com/kubeflow/kubeflow/issues/5302

Alternatively, for testing, you can use the `authservice_session` cookie for the Kubeflow dashboard site from the browser. Once
the content of the cookie is retrieved from the browser, it can be added as a header in your request

> (e.g. `"Cookie: authservice_session=MTYwNz..."`).

Reviewer 2: If the request is sending over CURL, it should use the header `X-Auth-Token: <token wtihout authservice_session from cookie> `. but cookie works, I won't block merge if using cookie header.

Author: Hmm, I can't seem to get that header to work.

This works:

```
curl -v -H "Cookie: authservice_session=MTYwODIzNDQyOHxOd3dBTkZoSk5VRTNNek5QUVRKR1NFTlNWRU5ITWpKRVJ6TTJWamRFVTA1RVNrbFhSelZOU0ZkUFZFeEtRMWRRVFRaS1RVeFNObEU9fPBUEE3vaC4PIxxHRwuphwFWf0DdbIDy-VA9_DQ2yuq5" https://sklearn-iris-bar-p-test.kfp-dev-442dbba0442be6c8c50f31ed96b00601-0001.sjc03.containers.appdomain.cloud/v1/models/sklearn-iris-bar:predict -d '{"instances": [[6.8,  2.8,  4.8,  1.4],[6.0,  3.4,  4.5,  1.6]]}'
```
But not this:
```
curl -v -H "X-Auth-Token: MTYwODIzNDQyOHxOd3dBTkZoSk5VRTNNek5QUVRKR1NFTlNWRU5ITWpKRVJ6TTJWamRFVTA1RVNrbFhSelZOU0ZkUFZFeEtRMWRRVFRaS1RVeFNObEU9fPBUEE3vaC4PIxxHRwuphwFWf0DdbIDy-VA9_DQ2yuq5" https://sklearn-iris-bar-p-test.kfp-dev-442dbba0442be6c8c50f31ed96b00601-0001.sjc03.containers.appdomain.cloud/v1/models/sklearn-iris-bar:predict -d '{"instances": [[6.8,  2.8,  4.8,  1.4],[6.0,  3.4,  4.5,  1.6]]}' 
```

Reviewer 2: I saw in Kubeflow v1.2, it still uses `cookie` header and `X-Auth-Token` header in the enovyfilter that forwards request to auth service: https://github.com/kubeflow/manifests/blob/b64b76b4ef3a984115f01686812588e81f46a02b/istio/oidc-authservice/base/envoy-filter.yaml#L19-L20

So `X-Auth-Token` should work, but anyway, won't block merge here.
